### ISSUE TITLE
verifying emails

### ISSUE THREAD

**Issue by ssun098** [keybase/keybase-issues#1299]
This is a feature request, but it would be great if you could add some sort of email verification! I think it would be very helpful with the authentication of certain people through email (aka scammers and such).

**Comment by dtiersch** [comment]
Verifications are public and public reviewable. This means that everybody has to be able to check the proofs for themselves without relying on any 3rd party services. (To check that the Github account belongs to you I only need the gist you posted and gpg).
Verifying emails would require that **everybody** needs to be able to send you an email and getting some kind of automated and signed reply to verify the ownership. This is not possible - especially without exposing your **unlocked** private key to a public accessible service.

**Comment by ssun098** [comment]
Oh that makes more sense.  Thanks for clarifying!

**Comment by zQueal** [comment]
Maybe we could make use of the keybase email forwarder to scrape verification emails and host them as text files under the keybase.io domain?

```
Delivered-To: zach.queal@gmail.com
Received: by 54.84.192.51 with SMTP id fu7csp5138051ieb;
        Tue, 6 Jan 2015 03:12:11 -0800 (PST)
X-Received: by 10.140.92.209 with SMTP id b75mr18652898qge.60.1420542730821;
        Tue, 06 Jan 2015 03:12:10 -0800 (PST)
Return-Path: <zqueal@keybase.io>
Received: from mail1.keybase.io (mail1.keybase.io. [54.84.192.51])
        by mx.google.com with ESMTP id v4si63753410qcz.47.2015.01.06.03.12.10
        for <zach.queal@gmail.com>;
        Tue, 06 Jan 2015 03:12:10 -0800 (PST)
Received: from keybase.io (keybase.io [54.84.192.51])
	by mail1.keybase.io (Postfix) with ESMTP id 1A3CB6055F
	for <zqueal@keybase.io>; Tue,  6 Jan 2015 11:12:10 +0000 (UTC)
Received: by keybase.io (Postfix, from userid 0)
	id C0A29140B94; Tue,  6 Jan 2015 06:12:09 -0500 (EST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple;
	d=keybase.io; s=mail; t=1420542729;
	bh=ZjWMiBr/W5xYhovrYBSne0B2gVHFGBQZSZQZcOAv50s=;
	h=Date:From:To:Subject:From;
	b=2dI7mDnsEbMFpV1ZwkeZTiAaMeNuL923l7a22TyYh11Cy2WN9u5ppNWa+87GGrSMk
	 joPJmCb4CX4Jf+WlN7VxeQzzOY7FX6O5LYTpc+8WUpm/t8ZoCpaNX+jWPWuI2wxXRj
	 bJGigi+vcxzncIY0PNRs0p/DGKSiMFRDIGodgFG4=
Date: Tue, 6 Jan 2015 06:12:09 -0500
From: Zach Queal <zqueal@keybase.io>
To: zqueal@keybase.io
Subject: Email Verification
Message-ID: <20150106111209.GA31323@keybase.io>
MIME-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline
User-Agent: Mutt/1.5.21 (2010-09-15)

-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

This email message will stand as proof that I own and operate the email address: zach.queal@gmail.com

My public key fingerprint is: 7B5AC0300E2C74FD355CCC81B1DA56DCB73E8516
-----BEGIN PGP SIGNATURE-----
Version: Keybase OpenPGP v2.0.1
Comment: https://keybase.io/crypto

wsFcBAABCgAGBQJUq8L4AAoJELHaVty3PoUW8ukP/2qxc4kyadrWMiQaqcu23anu
w0yd2Lge9kuThAANFeJ39btlZJJeTCXsan2L0M9VyXXiSwdgZTkp43T6kG3ssEPP
rpjc2Kj1CvGrdcQgwSm5yFV0zmqSLzy2UbCX/sPCQga+pX7ZB/B0hNw3VJMDFHDd
NLMGYwt9edO4B7w5OHENaluH6GpVf6PdlMIJqid3f0toivDkU6PyzBCQMXyr2lA4
PPT1hgwyUpOQ1lN1E/hYTdhcNNKq0zg0+Rvbiq8UPt6B+jKYqiFMUA/cMRnnRDzG
ZRzrgvrPNhUCn/9DEjnhfGrRolqoztS2b+agByZIKYRIgGZxmGfOjE4LlYF6/GyV
iAABGp17T+bx/zelcPL5XkQeuB3YJbS/3UhYW+AEtdj4ZYk05LUjxtxMxV/XJQu1
Esb1nFhQbFWTFMiA48VVtvFBUEGIn/KrXAV3k1KrSxrHujyCYeYB74AzY9fYzHTM
JYjKIGrSL/0aVYfDdWGgxYb45JGtWKTyoEG1gvFOugFVryBsK+UOhiu85cihnosi
oenG72jS0/eH5hVbGa+DCpGw3SOMKm694xTz00ebqFX32fPMRRPD8tUFpyz2RltW
tkVPtNFpZ16+J6PreVcxvCHdaZ77whdDHoUadX3AdwfblTJPvtZCxD2ofMJTqf24
s/wr+NPsAKqLM5EM3PqG
=AbeT
-----END PGP SIGNATURE-----
```

The user could then always verify the email contents and view the header information. It's a bit hacky, though.

**Comment by dtiersch** [comment]
that means you would have to *trust* a third party - in this case keybase.io - that the shown text is identical to the one originally received as email since you can mess with all header data as you wish. And you would have to copy by hand the complete email into keybase.io first.
This sounds neither public verifyable nor very trustworthy...

**Comment by malgorithms** [comment]
Yeah, as @dtiersch  said.

In the long run, all is not lost. There's been some discussion of DKIM (see #373) to prove emails. I still haven't read up on it, so no further comment.

Also, both Google and Yahoo are working on end-to-end encryption schemes, part of which involves them publishing a key directory. This works, too. If a GMail user gets a public key (whether or not it's the same one they use on Keybase), they can use their private to co-sign a statement with their Keybase-published key, making them agree. To verify you would trust Google's key server. This is the same amount of trust as trusting Github when you ask for the contents of a certain gist, or twitter, when you ask for a certain tweet. Google's key server could lie, just as Github or Twitter could lie. But it would have to lie in concert with Keybase.

In other words, we'd get the result we want: Keybase and the email provider would agree.

A solution where we just post an email puts all trust on Keybase, which is unacceptable. For the same reason we can't do OAuth type proofs  / private tweets.

**Comment by Fale** [comment]
Gmail users are also google plus users, so they can simply post the string on a google plus public post.

**Comment by GochoMugo** [comment]
@Fale good point since every gmail account is tied to a google plus account. But does that mean they are two accounts? Or just one? Confused :confused: but :+1:

**Comment by Fale** [comment]
@GochoMugo my understanding is that the account is a google account (at http://accounts.google.com) and gmail and plus are services allowed to use this (common) account.

**Comment by zQueal** [comment]
> good point since every gmail account is tied to a google plus account.

This is not entirely true. New gmail users have the option of **not** setting up a Google+ profile, although everyone has the option to. I just wanted to point that out--I have a few gmail accounts that do not have a Google+ profile.

**Comment by ssun098** [comment]
Then wouldn't a public key be viable for gmail users with Google+, even if that leaves out those without?

**Comment by dtiersch** [comment]
Can you get the G+ profile URL if you know only the email address? Otherwise such a link between Gmail and G+ wouldn't make any sense

**Comment by zQueal** [comment]
> Are you able to get the G+ profile URL if you know only the email address? Otherwise such a link between Gmail and G+ wouldn't make any sense

It may be possible some other way, however, I was not able to search for my email address in the Google+ search and return my user account, but I was able to use my name. Which is pretty useless. So I really don't think this is going to work out in our favor.

> Then wouldn't a public key be viable for gmail users with Google+, even if that leaves out those without?

I'm sure there's a way. Google APIs are usually very complete. Only time will tell, though.

**Comment by briantopping** [comment]
Why all this necessary if I can simply add emails to the key on my client and do a ```keybase push --update```? Since I have my privkey and the act of adding an email attribute to my pubkey relies on me having the privkey, it follows that I am the only one that could do this (or my key is compromised, different issue). 

When the pubkey itself is carrying around the email addresses, if I verify the key, the emails are de-facto verified.

**Comment by zQueal** [comment]
> it follows that I am the only one that could do this

Well, the way that I was always taught is that you should always operate under the assumption that any one piece of identity information is compromised. Which is why identity profiles are pretty great as you have more identity avenues to ensure authenticity. So it really means nothing to me, as a third party that can't truly identify your identity that you can add an email to your key. That doesn't really give any indication other than _someone_ has your private key and has added an email to it--by extension, if I can't trust that _you_ and _only you_ own your private key, what use is there in trusting the emails as well?

I might be __entirely__ off base here, which has been known to happen on more than one occasion, but I'm pretty sure this is why the WOT was formed, and is so vital to the PGP community.

**Comment by briantopping** [comment]
@zQueal: Maybe I am the one missing something here.. If the integrity of the certificate (pubkey) is intact, how does an attacker insert a bogus email address? Even if they could, what value would it be if they didn't have the private key to decrypt the email when it arrived?

Your point about the role of the identity profile seems correct to me, that I can use the profile to get much greater confidence that a cert is from who I think it is from instead a bunch of bits that happen to be well-formed from a site that looks like Craig's List. Once I accept a certificate as authentic and am 100% sure that the certificate was installed on my machine without tampering, the identity profile falls away. 

It's more of a coincidence to me that Keybase offers encryption or signing, it doesn't actually provide unique value.

Emails addresses in the key are endpoint addresses to reach the owner. In contrast to proof URL addresses or social media addresses a cert owner may be on, email addresses are writable endpoints. IMHO, only writable endpoints make sense as endpoint attributes in a key. Read only attributes do not provide long term value once the key is accepted as authentic. Once downloaded, even if my MUA is compromised, the keystore can be cryptographically validated in an open manner every time it is used. Thus, I want the endpoint addresses to be a part of this validation, not in a lookaside table that was constructed based on a conversation with an external server (that table is proprietary to the MUA and can't be validated in an open manner).

Lastly, most cryptographically enhanced MUAs will refuse to encrypt a mail using a certificate that does not contain the recipient address. So it needs to be in there anyway.

Long story short, if I trust the stream of bits that make up a certificate via an identity profile and the certificate has email addresses of the user, I hold that the email addresses are both verified and secure without any external system like is proposed here.

**Comment by Fale** [comment]
In reality we can not completely trust the public key, because I can add the email obama@whitehouse.gov without having it and I have no proof to bring before adding it to the public key, while I should if I was going to add it in keybase...

**Comment by briantopping** [comment]
@Fale: Isn't that what ```keybase verify``` is supposed to do? If not, shouldn't it?

When I first started the other day, I uploaded my public certificate from pgp.mit.edu and it wanted me to verify the certificate before I could use the account. I would have thought the verification would have failed if I didn't have the private key and didn't even try it.

**Comment by Fale** [comment]
@briantopping I'm referring to the phrases

"Since I have my privkey and the act of adding an email attribute to my pubkey relies on me having the privkey, it follows that I am the only one that could do this (or my key is compromised, different issue)."
This is indeed true.

"When the pubkey itself is carrying around the email addresses, if I verify the key, the emails are de-facto verified."
This is false. Keybase asks you to verify one email, not each one. So if I have my true mail and a fake mail, we can not assume that since the user has passed the validation all emails are ok.
For instance I could have 3 fake emails and a real one and still passing the keybase verify test... http://keys.fedoraproject.org/pks/lookup?op=index&search=0x6D6A29D6709A7851

**Comment by briantopping** [comment]
@Fale: Yes, I see what you are referring to now. I am assuming that I would not be an idiot and put the wrong email addresses in my own key.

I would like to propose that this doesn't matter though. Say I add obama@whitehouse.gov to my list of email addresses. People send "Hey Brian, how are you here's tons of bitcoin attached blah blah"  to me at the wrong address, hopefully encrypted with a key that the recipient can't do anything with the data anyway. I'm an idiot, people sending me emails that I never get, but no damage done because they can't read them.

But if someone sends same email without a key to obama@wh.gov and I am not there to get it, what is the difference between those situations? I still didn't get the email, the sender is an idiot in that case. This is the same without GPG. Nobody does this.

Third case is someone wants to send me bitcoin and they don't know my cert has been tampered with, all email addresses changed to attacker's email addresses, fingerprint doesn't match server. Does it matter attacker address got encrypted value? They can't do anything with it since they don't have the privkey, but really the cert fingerprint should have been checked first, no?

I guess I am just confused why it matters if the endpoint address is false if the cert is provably tampered with. If the content is in a hardened envelope, should be an annoyance more than anything.

Would it be better to focus on verifying keys than verifying endpoints?

**Comment by Fale** [comment]
@briantopping I can see partially your point.

The attack vector for this case is:

* I make a lot of people signing my key (with my real name and my real email)
* I put an obama@wh.gov email to my key
* I rename my key to "Barack Obama"
* I send an email that seems to arrive from obama@wh.gov duly signed (with my key) requesting to send bitcoin to my bitcoin (my = the attacker)
* The receiver does receive the mail that _seems_ to arrive from Obama and the GPG key checks out.
* The receiver sends money

Clearly, Obama is not the best person for this kind of scam, but other people could.

**Comment by briantopping** [comment]
I accept this as problem. It might be solved if certain changes like name change caused my key to require re-watching by all watchers. If a friend gets married and changes her name, I would be good with being asked to confirm it. OTOH if her account got hacked, I wouldn't want hackers doing things with it and would not confirm. The difference is obvious to people that know each other and if I don't know the difference about someone, I probably shouldn't track them anyway. Lack of confirm means lack of track. 

I am followed every day by people on Twitter that are obviously building a reputation there for some nefarious purpose and want me to follow them back. Sometimes their tweets are funny and people find safety in the number of followers so they don't ask what harm they might be eventually helping. What you describe could easily happen in such a case. TaylorSwiftInfoSec changes name to Barak Obama etc. 

It's important to protect the Keybase brand. I don't think Keybase should publish a cert that doesn't have a @keybase.io email in it and if users rely on Keybase without monitoring fingerprints, there may be nothing that can be done to protect those users.

**Comment by zQueal** [comment]
> If the integrity of the certificate (pubkey) is intact, how does an attacker insert a bogus email address? Even if they could, what value would it be if they didn't have the private key to decrypt the email when it arrived?

This is a bit oxymoronical. If they're able to insert invalid email addresses into the key, they already have the private key, so they would certainly be able to decrypt any message encrypted with the public key. 

The entire issue with PGP and trust, is that the system isn't perfect. Even with a totally valid PGP key, certified by 100 or more people theres still a possibility that the person sending you information encrypted with a public key still isn't the person they claim to be. PGP has always been a lowest common denominator encryption method; which is to say that no matter what you do you're still taking a risk. There are a few reasons for this. Most of which involve the care that people take with their private key. There are some that don't keep it in liquid storage, which is great. But I've also seen people upload them to private web servers that are indexable by Google. It can be a disaster if you let it. Based on this assumption the validity of keys, even those that are certified, IMHO, decrease over time. Of course this is just all my opinion, and like I said I could be completely out of line, and I'm still not saying that you're wrong. I'm simply saying that just because someone can add an email address to a key doesn't mean that it should be trusted. There's no public way to verify that the email is valid; unlike publicly auditable proofs, like Twitter, where I can go any time and test to see if your key is still verifying that it owns it. For that to be spoofed, not only does someone have to have your private key, but they need to know the password (hopefully you've chosen a strong one) as well as the password to your Twitter account to post the proof. Even if they do have all of that, you can simply recover your account, remove the proof, and it's no longer valid. With adding an email address to a key, I can't do that. I have to explicitly trust that a person is exactly who they say they are.

If that makes any sense.

**Comment by briantopping** [comment]
I thought someone here was saying that one could insert email addresses without having the private key. I give up, sincere apologies for the interruption....

**Comment by jokeyrhyme** [comment]
So, GitHub itself now supports GPG signature verification:
https://github.com/blog/2144-gpg-signature-verification

However, I suspect that it is sending verification emails to jokeyrhyme@keybase.io (given that my profile is https://keybase.io/jokeyrhyme ).

Is there a way to have these emails forwarded through to my real (and private) email address?

**Comment by strib** [comment]
@jokeyrhyme: see https://github.com/keybase/keybase-issues/issues/245#issuecomment-205961692

**Comment by jokeyrhyme** [comment]
For reference, I followed this guide: https://www.katescomment.com/how-to-add-additional-email-addresses-to-your-gpg-identity/
Then I ran `keybase push --update`.
Then I copied the new public key over to GitHub. Everything is fine now. :D
